Lib/ast.py

```diff
@@ -1236,17 +1236,35 @@ def visit_JoinedStr(self, node):
 
         new_fstring_parts = []
         quote_types = list(_ALL_QUOTES)
+        fallback_to_repr = False
         for value, is_constant in fstring_parts:
             if is_constant:
-                value, quote_types = self._str_literal_helper(
+                value, new_quote_types = self._str_literal_helper(
                     value,
                     quote_types=quote_types,
                     escape_special_whitespace=True,
                 )
+                if set(new_quote_types).isdisjoint(quote_types):
+                    fallback_to_repr = True
+                    break
+                quote_types = new_quote_types
             elif "\n" in value:
                 quote_types = [q for q in quote_types if q in _MULTI_QUOTES]
```

Comment:
Should we also add a check here?
```suggestion
                quote_types = [q for q in quote_types if q in _MULTI_QUOTES]
                if not quote_types:
                    fallback_to_repr = True
                    break
```

Reply:
Hm, I think this is unreachable. In general, if we rule out a multiquote, we also have ruled out its single quote. The only case where `_str_literal_helper` does something different is the disjoint repr case (note it searches in `quote_types` first). I'll add an assert and more test cases though.